modules/text/include/opencv2/text/textDetector.hpp

```diff
@@ -0,0 +1,271 @@
+// any express or implied warranties, including, but not limited to, the implied
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#ifndef __OPENCV_TEXT_TEXTDETECTOR_HPP__
+#define __OPENCV_TEXT_TEXTDETECTOR_HPP__
+
+#include <vector>
+#include <string>
+#include <iostream>
+#include <sstream>
+#include"ocr.hpp"
+
+
+namespace cv
+{
+namespace text
+{
+
+//! @addtogroup text_detect
+//! @{
+
+
+
+//base class BaseDetector declares a common API that would be used in a typical text
+//detection scenario
+class CV_EXPORTS_W BaseDetector
+{
+public:
+    virtual ~BaseDetector() {};
+
+    virtual void run(Mat& image,
+                     std::vector<Rect>* component_rects=NULL,
+                     std::vector<float>* component_confidences=NULL,
+                     int component_level=0) = 0;
+
+    virtual void run(Mat& image, Mat& mask,
+                     std::vector<Rect>* component_rects=NULL,
+                     std::vector<float>* component_confidences=NULL,
+                     int component_level=0) = 0;
+
+};
+/** A virtual class for different models of text detection (including CNN based deep models)
+ */
+
+class CV_EXPORTS_W TextRegionDetector
+{
+protected:
+    /** Stores input and output size
+     */
+    //netGeometry inputGeometry_;
+    //netGeometry outputGeometry_;
+    Size inputGeometry_;
+    Size outputGeometry_;
+    int inputChannelCount_;
+    int outputChannelCount_;
+
+public:
+    virtual ~TextRegionDetector() {}
+
+    /** @brief produces a list of Bounding boxes and an estimate of text-ness confidence of Bounding Boxes
+     */
+    CV_WRAP virtual void detect(InputArray image, OutputArray bboxProb ) = 0;
+
+
+    /** @brief simple getter method returning the size (height, width) of the input sample
+     */
+    CV_WRAP virtual Size  getInputGeometry(){return this->inputGeometry_;}
+
+    /** @brief simple getter method returning the shape of the oputput
+     *   Any text detector should output a number of text regions alongwith a score of text-ness
+     *   From the shape it can be inferred the number of text regions and number of returned value
+     *   for each region
+     */
+    CV_WRAP virtual Size getOutputGeometry(){return this->outputGeometry_;}
+
+
+
+};
+
+/** Generic structure of Deep CNN based Text Detectors
+ * */
+class CV_EXPORTS_W  DeepCNNTextDetector : public TextRegionDetector
```

Comment:
So, what is the typical way to use the `DeepCNNTextDetector`? Is it to call `create()` and then call detect? Or is this some intermediate class that gets inherited by some other class?

Reply:
DeepCNNTextDetector is created as a generic CNN based text detector, it inherits TextRegionDetector, which can be any text detector. DeepCNNTextDetector implements detect function to provide bounding boxes of text.